arroyo/backends/kafka/consumer.py

```diff
@@ -246,40 +252,74 @@ def assignment_callback(
+                try:
+                    incremental_assignment: MutableSequence[
+                        ConfluentTopicPartition
+                    ] = []
+
+                    for partition in partitions:
+                        if partition.offset >= 0:
+                            incremental_assignment.append(partition)
+                        elif partition.offset == OFFSET_INVALID:
+                            incremental_assignment.append(
+                                self.__resolve_partition_starting_offset(partition)
+                            )
+                        else:
+                            raise ValueError("received unexpected offset")
+
+                    offsets = {
+                        Partition(Topic(i.topic), i.partition): i.offset
+                        for i in incremental_assignment
+                    }
+
+                    self.__incremental_assign(offsets)
+
+                    # Ensure that all partitions are resumed on assignment to avoid
+                    # carrying over state from a previous assignment.
+                    self.resume([p for p in offsets])
+
+                except Exception:
+                    self.__state = KafkaConsumerState.ERROR
+                    raise
 
-                for partition in offsets:
-                    self.__paused.discard(partition)
-            except Exception:
-                self.__state = KafkaConsumerState.ERROR
-                raise
+            else:
+                try:
+                    assignment: MutableSequence[ConfluentTopicPartition] = []
+
+                    for partition in self.__consumer.committed(partitions):
+                        if partition.offset >= 0:
+                            assignment.append(partition)
+                        elif partition.offset == OFFSET_INVALID:
+                            assignment.append(
+                                self.__resolve_partition_starting_offset(partition)
+                            )
+                        else:
+                            raise ValueError("received unexpected offset")
+
+                    offsets = {
+                        Partition(Topic(i.topic), i.partition): i.offset
+                        for i in assignment
+                    }
+
+                    self.__assign(offsets)
+
+                    # Ensure that all partitions are resumed on assignment to avoid
+                    # carrying over state from a previous assignment.
+                    self.__consumer.resume(
+                        [
+                            ConfluentTopicPartition(
+                                partition.topic.name, partition.index, offset
+                            )
+                            for partition, offset in offsets.items()
+                        ]
+                    )
+
+                    for partition in offsets:
+                        self.__paused.discard(partition)
```

Comment:
Any reason for relying on the `resume` method for the incremental method and, instead, reimplementing its logic here for the stop-the-world rebalancing ?

Reply:
No reason, I just mostly avoided touching this implementation (it's just indented differently so it shows up here). But yes it's cleaner to reuse `resume` everywhere, I'll switch it over.